File: js/path-crossing.js
export const isPathCrossing = function (path) {
  const visited = new Set();
  let x = 0;
  let y = 0;
  visited.add("0:0");

  for (let i = 0; i < path.length; i++) {
    const it = path.at(i);

    switch (it) {
      case "N":
        y--;
        break;
      case "S":
        y++;
        break;
      case "W":
        x--;
        break;
      case "E":
        x++;
        break;
    }
    const key = x + ":" + y;

    if (visited.has(key)) {
      return true;
    } else {
      visited.add(key);
    }
  }

  return false;
};

File: go/path-crossing.go
package main

import "fmt"

func isPathCrossing(path string) bool {
  visited := make(map[string]struct{})
  x, y := 0, 0
  visited["0:0"] = struct{}{}

  for _, it := range path {
    switch it {
      case 'N':
        y--
      case 'S':
        y++
      case 'W':
        x--
      case 'E':
        x++
    }

    key := fmt.Sprintf("%d:%d", x, y)

    if _, found := visited[key]; found {
      return true
    } else {
      visited[key] = struct{}{}
    }
  }
  return false
}
